### PR TITLE
only add direct build-only dependencies to PATH

### PR DESCRIPTION
**Pull request by davydden**
fixes https://github.com/LLNL/spack/issues/4064

- [x] @hartzell please see if this fixes the  problem
- [ ] add tests

### CODE REVIEW

File: lib/spack/spack/build_environment.py [hunk 1]
```diff
@@ -248,7 +248,7 @@ def set_build_environment_variables(pkg, env, dirty=False):
         dirty (bool): Skip unsetting the user's environment settings
     """
     # Gather information about various types of dependencies
-    build_deps      = pkg.spec.traverse(root=False, deptype=('build'))
+    build_deps      = pkg.spec.dependencies(deptype='build')
```
Comment (eschnett) [inline lib/spack/spack/build_environment.py]: Shouldn't run-time dependencies of direct build-time dependencies also be added?

Comment (davydden) [inline lib/spack/spack/build_environment.py]: probably

**Comment by davydden** [comment]
within this patch `spack env gsmh bash` contains:
```
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/openblas-0.2.19-c2gdrjmns3xqq7wmlau6ttkhblfhwje3/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/cmake-3.7.2-x2vsvmeg76dkdsslny3rhgwk66fej5nh/bin
```
without
```
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/sqlite-3.18.0-5vkemqqym4maw6r73zzlia4msh62jr5j/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/readline-7.0-mnwyhyw3kef5ndyxzamz2s6w5jd7o54h/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/bzip2-1.0.6-k42bho7lywu6laeutlfh7q3mivn3h3qk/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/python-2.7.13-lvkcqfenrdhqjb43ok5hezmosam6gfab/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/metis-5.1.0-qcvico4siizb4xlbrkkboulhfqhbz7gw/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/xz-5.2.3-5xaogvjybobt4kmdgeyv5nfpclfmjgqa/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/libxml2-2.9.4-nniu7gtxm2ijknfj6brv4a74goxhaszz/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/hwloc-1.11.6-pkbyijayr66g3wq3hojj3l44qc7kjno3/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/openblas-0.2.19-c2gdrjmns3xqq7wmlau6ttkhblfhwje3/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/m4-1.4.18-4f4cjgw5fcggj65qk3tzuhkccgv4prn2/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/openssl-1.0.2k-v54rcq3ftv4vg7ss4uloaulfh2ytds36/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/pkg-config-0.29.2-72rz43flet5qcqtnx6zimnqy4ilk3tys/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/ncurses-6.0-kcgagcxwgv5jtbr3omi3v2nimkypq2nf/bin:
/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/cmake-3.7.2-x2vsvmeg76dkdsslny3rhgwk66fej5nh/bin
```

File: lib/spack/spack/build_environment.py (hunk 1 above)
Comment (davydden) [inline lib/spack/spack/build_environment.py]: i added this.

**Comment by davydden** [comment]
@tgamblin where `set_build_environment_variables()` is tested? I would like to add a few tests on `PATH` build environment variable to this PR.

**Comment by hartzell** [comment]
I have a build running.

**Comment by hartzell** [comment]
This worked for me, details in #4064 (sorry, split brain...).

**Comment by hartzell** [comment]
@davydden -- are we waiting for a response from @tgamblin re testing `set_build_environment_variables`?   I'd like to get my CI run blue again, either waiting for this merge or skipping a couple of bits.

Sniffing around with an innocent's eye, I see that `MakeExecutable`, also defined in `build_environment.py` has it's own test file, `.../lib/spack/spack/test/make_executable.py`.

**Comment by davydden** [comment]
@hartzell I am ok to merge it as-is to fix the bug ASAP.   @tgamblin @adamjstewart thoughts?

**Pull request merged**